src/rgw/rgw_pubsub.cc

```diff
@@ -462,13 +475,73 @@ std::string rgw_pubsub_dest::to_json_str() const
   return ss.str();
 }
 
+void rgw_pubsub_dest::decode_json(JSONObj* f) {
+  using rgw::notify::DEFAULT_CONFIG;
+  using rgw::notify::DEFAULT_GLOBAL_VALUE;
+  JSONDecoder::decode_json("push_endpoint", push_endpoint, f);
+  JSONDecoder::decode_json("push_endpoint_args", push_endpoint_args, f);
+  JSONDecoder::decode_json("push_endpoint_topic", arn_topic, f);
+  JSONDecoder::decode_json("stored_secret", stored_secret, f);
+  JSONDecoder::decode_json("persistent", persistent, f);
+  std::string ttl;
+  JSONDecoder::decode_json("time_to_live", ttl, f);
+  time_to_live = ttl == DEFAULT_CONFIG ? DEFAULT_GLOBAL_VALUE : std::stoul(ttl);
+
+  std::string max_retry;
+  JSONDecoder::decode_json("max_retries", max_retry, f);
+  max_retries = max_retry == DEFAULT_CONFIG ? DEFAULT_GLOBAL_VALUE
+                                            : std::stoul(max_retry);
+
+  std::string sleep_dur;
+  JSONDecoder::decode_json("retry_sleep_duration", sleep_dur, f);
+  retry_sleep_duration = sleep_dur == DEFAULT_CONFIG ? DEFAULT_GLOBAL_VALUE
+                                                     : std::stoul(sleep_dur);
+}
+
 RGWPubSub::RGWPubSub(rgw::sal::Driver* _driver, const std::string& _tenant)
   : driver(_driver), tenant(_tenant)
 {}
 
-int RGWPubSub::read_topics(const DoutPrefixProvider *dpp, rgw_pubsub_topics& result, 
+int RGWPubSub::read_topics(const DoutPrefixProvider *dpp, rgw_pubsub_topics& result,
     RGWObjVersionTracker *objv_tracker, optional_yield y) const
 {
+  if (driver->get_zone()->get_zonegroup().supports_feature(
+          rgw::zone_features::notification_v2)) {
```

Comment:
thinking about multi-zonegroup realms, i think all of these checks need to require that all zonegroups in the realm have enabled this

Reply:
> thinking about multi-zonegroup realms, i think all of these checks need to require that all zonegroups in the realm have enabled this

you mean, a for loop to check for all the zongroups ?
is there a quick API, that gives you the list of zonegroups ? `sysobj_svc->get_zone_svc()->list_zonegroups`

Reply:
ultimately this would come from the `RGWPeriod` which contains `RGWPeriodMap` with a map of all `RGWZoneGroups` in the realm. you can get to this either through `RGWSI_Zone::get_current_period()` if you're behind the rados store, or from `rgw::SiteConfig::get_period()` otherwise

Reply:
> ultimately this would come from the `RGWPeriod` which contains `RGWPeriodMap` with a map of all `RGWZoneGroups` in the realm. you can get to this either through `RGWSI_Zone::get_current_period()` if you're behind the rados store, or from `rgw::SiteConfig::get_period()` otherwise

Thanks, i saw that interface and to get the period we need to static cast the `driver`to `rgw_pubsub.cc` files which previously was flagged. So is it okie to do that?
But why do all the zonegroup have to have the feature enabled ? syncing is between zones in a give zonegroup ? not across zones in different zonegroup ?
`

Reply:
> But why do all the zonegroup have to have the feature enabled ? syncing is between zones in a give zonegroup ? not across zones in different zonegroup ?

data sync is only between zones in a zonegroup, but for metadata sync all zones in the entire realm sync from the metadata master zone

the `resharding` feature only applies to data sync, so we only check the local zonegroup for that one

Reply:
> Thanks, i saw that interface and to get the period we need to static cast the `driver`to `rgw_pubsub.cc` files which previously was flagged. So is it okie to do that?

callers from rgw_rest_pubsub.cc would have access to the SiteConfig through `s->penv.site`, so could pass it in. i guess it's the uses in rgw_admin.cc that don't have one? i'm trying to get @adamemerson's https://github.com/ceph/ceph/pull/52496 merged soon that adds a SiteConfig to rgw_admin.cc. should we follow up on this after that one merges?

i'd say that casts to RadosStore should be a last resort, because they won't work with other backends, or if there are filters like posix/d3n/d4n in front of rados

Reply:
> > Thanks, i saw that interface and to get the period we need to static cast the `driver`to `rgw_pubsub.cc` files which previously was flagged. So is it okie to do that?
> 
> callers from rgw_rest_pubsub.cc would have access to the SiteConfig through `s->penv.site`, so could pass it in. i guess it's the uses in rgw_admin.cc that don't have one? i'm trying to get @adamemerson's #52496 merged soon that adds a SiteConfig to rgw_admin.cc. should we follow up on this after that one merges?
> 
> i'd say that casts to RadosStore should be a last resort, because they won't work with other backends, or if there are filters like posix/d3n/d4n in front of rados

so `rgw_admin.cc`, `rgw_pubsub.cc` and `driver/rados/rgw_bucket.cc` are 3 places where we do not have direct access to RgwPeriod, `rgw_admin` would have access SiteConfig after #52496, but the other 2 places will still have problem

Reply:
> so `rgw_admin.cc`, `rgw_pubsub.cc` and `driver/rados/rgw_bucket.cc` are 3 places where we do not have direct access to RgwPeriod, `rgw_admin` would have access SiteConfig after #52496, but the other 2 places will still have problem

a lot of the calls to rgw_pubsub.cc come from rgw_rest_pubsub.cc, which would have access through `s->penv.site`. the `bucket_stats()` case in driver/rados/rgw_bucket.cc should either be called from rgw_admin.cc or from driver/rados/rgw_rest_bucket.cc which has `s->penv.site`

though i don't think `bucket_stats()` should be dumping stuff about notification topics; we already have `radosgw-admin notification list --bucket=` and `RGWPSListNotifsOp` for that

Reply:
> > so `rgw_admin.cc`, `rgw_pubsub.cc` and `driver/rados/rgw_bucket.cc` are 3 places where we do not have direct access to RgwPeriod, `rgw_admin` would have access SiteConfig after #52496, but the other 2 places will still have problem
> 
> a lot of the calls to rgw_pubsub.cc come from rgw_rest_pubsub.cc, which would have access through `s->penv.site`. the `bucket_stats()` case in driver/rados/rgw_bucket.cc should either be called from rgw_admin.cc or from driver/rados/rgw_rest_bucket.cc which has `s->penv.site`
> 
> though i don't think `bucket_stats()` should be dumping stuff about notification topics; we already have `radosgw-admin notification list --bucket=` and `RGWPSListNotifsOp` for that

agree, i could remove  the dumping of notification from `bucket stats`
regarding having access to `s->penv.site`, you mean we pass `s->penv.site` to `get_topics`/`get_topic`/`create_topic`/`remove_topic` ? the api's are invoked via `radosgw-admin` or via `s3`, so you want the pubusub class to take  `s->penv.site` along with tenant and driver and decide wether to use v2 or v1?

Reply:
> so you want the pubusub class to take `s->penv.site` along with tenant and driver and decide wether to use v2 or v1?

that sounds right, yeah. https://github.com/ceph/ceph/pull/52496#issuecomment-1904081498 isn't quite ready yet though

Reply:
So then we wait until #52496 is ready ?
Or do that in separate PR ?